src/Controls/src/Core/SetterSpecificity.cs

```diff
@@ -0,0 +1,96 @@
+
+		public static readonly SetterSpecificity ManualValueSetter = new SetterSpecificity(0, 100, 0, 0, 0, 0, 0, 0);
+		public static readonly SetterSpecificity Trigger = new SetterSpecificity(0, 101, 0, 0, 0, 0, 0, 0);
+
+		public static readonly SetterSpecificity DynamicResourceSetter = new SetterSpecificity(0, 0, 1, 0, 0, 0, 0, 0);
+
+		//handler always apply, but are removed when anything else comes in. see SetValueActual
+		public static readonly SetterSpecificity FromHandler = new SetterSpecificity(1000, 0, 0, 0, 0, 0, 0, 0);
+
+		//100-n: direct VSM (not from Style), n = max(99, distance between the RD and the target)
+		public int Vsm { get; }
+
+		//1: SetValue, SetBinding
+		public int Manual { get; }
+
+		//1: DynamicResource
+		public int DynamicResource { get; }
+
+		//1: SetBinding
+		public int Binding { get; }
+
+		//XAML Style specificity
+		//100-n: implicit style, n = max(99, distance between the RD and the target)
+		//200-n: RD Style, n = max(99, distance between the RD and the target)
+		//200: local style, inline css,
+		//300-n: VSM, n = max(99, distance between the RD and the target)
+		//300: !important (not implemented)
+		public int Style { get; }
+
+		//CSS Specificity, see https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity
+		public int Id { get; }
+		public int Class { get; }
+		public int Type { get; }
+
+		public SetterSpecificity(int vsm, int manual, int dynamicresource, int binding, int style, int id, int @class, int type)
+		{
+			Vsm = vsm;
+			Manual = manual;
+			DynamicResource = dynamicresource;
+			Binding = binding;
+			Style = style;
+			Id = id;
+			Class = @class;
+			Type = type;
+		}
+
+		public SetterSpecificity(int style, int id, int @class, int type) : this(0, 0, 0, 0, style, id, @class, type)
+		{
+		}
+
+		public int CompareTo(SetterSpecificity other)
+		{
+			//everything coming from Style has lower priority than something that does not
+			if (Style != other.Style && Style == 0)
+				return 1;
+			if (Style != other.Style && other.Style == 0)
+				return -1;
+			if (Style != other.Style)
+				return Style.CompareTo(other.Style);
+
+			if (Vsm != other.Vsm)
+				return Vsm.CompareTo(other.Vsm);
+			if (Manual != other.Manual)
+				return Manual.CompareTo(other.Manual);
+			if (DynamicResource != other.DynamicResource)
+				return DynamicResource.CompareTo(other.DynamicResource);
+			if (Binding != other.Binding)
+				return Binding.CompareTo(other.Binding);
+			if (Id != other.Id)
+				return Id.CompareTo(other.Id);
+			if (Class != other.Class)
+				return Class.CompareTo(other.Class);
+			return Type.CompareTo(other.Type);
+		}
+
+		public override bool Equals(object obj) => Equals((SetterSpecificity)obj);
+
+		bool Equals(SetterSpecificity other) => CompareTo(other) == 0;
+
+		public override int GetHashCode() => (Vsm, Manual, DynamicResource, Binding, Style, Id, Class, Type).GetHashCode();
```

Comment:
it returns 0 for the non-generic ValueTuple. using a tuple to write GetHashCode implementation is an encouraged pattern since c#7 (unless your tuple has more than 8 members, in that case, well, it depends.

actual implementation for generic value tuple GetHashCode 
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/ValueTuple.cs,2005

Reply:
FYI, today I found `HashCode.Combine()` we could use for cases like this in the future:

* https://github.com/xamarin/xamarin-macios/issues/16677
* https://github.com/dotnet/runtime/issues/9433